File: src/mcp/tools/definitions/foreshadowing_create.ts
/**
 * foreshadowing_createツール定義
 * 伏線要素を作成するMCPツール
 */

import type {
  McpToolDefinition,
  ToolExecutionContext,
} from "../tool_registry.ts";
import { executeCliCommand } from "../cli_adapter.ts";
import { ElementForeshadowingCommand } from "../../../cli/modules/element/foreshadowing.ts";

export const foreshadowingCreateTool: McpToolDefinition = {
  name: "foreshadowing_create",
  description:
    "伏線要素を作成します。物語の伏線（チェーホフの銃、予言、謎など）を管理するための構造を定義します。",
  inputSchema: {
    type: "object",
    properties: {
      name: {
        type: "string",
        description: "伏線名",
      },
      type: {
        type: "string",
        enum: [
          "hint",
          "prophecy",
          "mystery",
          "symbol",
          "chekhov",
          "red_herring",
        ],
        description:
          "伏線の種類: hint（ヒント）, prophecy（予言）, mystery（謎）, symbol（象徴）, chekhov（チェーホフの銃）, red_herring（レッドヘリング）",
      },
      plantingChapter: {
        type: "string",
        description: "伏線が設置されるチャプターID",
      },
      plantingDescription: {
        type: "string",
        description: "伏線設置の説明",
      },
      summary: {
        type: "string",
        description: "伏線の概要",
      },
      id: {
        type: "string",
        description: "伏線ID（省略時はnameから自動生成）",
      },
      importance: {
        type: "string",
        enum: ["major", "minor", "subtle"],
        description: "重要度: major（主要）, minor（副次）, subtle（微妙）",
      },
      plannedResolutionChapter: {
        type: "string",
        description: "回収予定チャプターID",
      },
      relatedCharacters: {
        type: "array",
        items: { type: "string" },
        description: "関連キャラクターID",
      },
      relatedSettings: {
        type: "array",
        items: { type: "string" },
        description: "関連設定ID",
      },
      displayNames: {
        type: "array",
        items: { type: "string" },
        description: "表示名のバリエーション",
      },
    },
    required: ["name", "type", "plantingChapter", "plantingDescription"],
  },
  execute: async (
    args: Record<string, unknown>,
    context?: ToolExecutionContext,
  ) => {
    const name = args.name as string | undefined;
    const type = args.type as string | undefined;
    const plantingChapter = args.plantingChapter as string | undefined;
    const plantingDescription = args.plantingDescription as string | undefined;

    if (!name || typeof name !== "string" || name.trim() === "") {
      return {
        content: [
          {
            type: "text" as const,
            text: "Error: 'name' parameter is required.",
          },
        ],
        isError: true,
      };
    }

    if (!type || typeof type !== "string") {
      return {
        content: [
          {
            type: "text" as const,
            text:
              "Error: 'type' parameter is required (hint|prophecy|mystery|symbol|chekhov|red_herring).",
          },
        ],
        isError: true,
      };
    }

    const validTypes = [
      "hint",
      "prophecy",
      "mystery",
      "symbol",
      "chekhov",
      "red_herring",
    ];
    if (!validTypes.includes(type)) {
      return {
        content: [
          {
            type: "text" as const,
            text: `Error: Invalid type '${type}'. Must be one of: ${
              validTypes.join(", ")
            }.`,
          },
        ],
        isError: true,
      };
    }

    if (
      !plantingChapter ||
      typeof plantingChapter !== "string" ||
      plantingChapter.trim() === ""
    ) {
      return {
        content: [
          {
            type: "text" as const,
            text: "Error: 'plantingChapter' parameter is required.",
          },
        ],
        isError: true,
      };
    }

    if (
      !plantingDescription ||
      typeof plantingDescription !== "string" ||
      plantingDescription.trim() === ""
    ) {
      return {
        content: [
          {
            type: "text" as const,
            text: "Error: 'plantingDescription' parameter is required.",
          },
        ],
        isError: true,
      };
    }

    // CLIコマンド用の引数を構築
    const commandArgs: Record<string, unknown> = {
      name,
      type,
      "planting-chapter": plantingChapter,
      "planting-description": plantingDescription,
    };

    if (typeof args.summary === "string") {
      commandArgs.summary = args.summary;
    }

    if (typeof args.id === "string") {
      commandArgs.id = args.id;
    }

    if (typeof args.importance === "string") {
      commandArgs.importance = args.importance;
    }

    if (typeof args.plannedResolutionChapter === "string") {
      commandArgs["planned-resolution-chapter"] = args.plannedResolutionChapter;
    }

    if (Array.isArray(args.relatedCharacters)) {
      const chars = (args.relatedCharacters as unknown[])
        .filter((c) => typeof c === "string")
        .map((c) => (c as string).trim())
        .filter((c) => c.length > 0);
      if (chars.length > 0) {
        commandArgs["related-characters"] = chars.join(",");
      }
    }

    if (Array.isArray(args.relatedSettings)) {
      const settings = (args.relatedSettings as unknown[])
        .filter((s) => typeof s === "string")
        .map((s) => (s as string).trim())
        .filter((s) => s.length > 0);
      if (settings.length > 0) {
        commandArgs["related-settings"] = settings.join(",");
      }
    }

    if (Array.isArray(args.displayNames)) {
      const displayNames = (args.displayNames as unknown[])
        .filter((n) => typeof n === "string")
        .map((n) => (n as string).trim())
        .filter((n) => n.length > 0);
      if (displayNames.length > 0) {
        commandArgs.displayNames = displayNames.join(",");
      }
    } else if (typeof args.displayNames === "string") {
      commandArgs.displayNames = args.displayNames;
    }

    const handler = new ElementForeshadowingCommand();
    return await executeCliCommand(handler, commandArgs, context?.projectRoot);
  },
};

File: internal/mcp/tools/foreshadowing_create.go
package tools

import (
	"context"
	"encoding/json"
	"fmt"
	"regexp"
	"strings"

	"github.com/takets/street-storyteller/internal/domain"
	"github.com/takets/street-storyteller/internal/mcp/protocol"
)

// ForeshadowingCreateTool creates a new foreshadowing entity and reports its ID.
type ForeshadowingCreateTool struct{}

type foreshadowingCreateArgs struct {
	ID                       string `json:"id"`
	Name                     string `json:"name"`
	Type                     string `json:"type"`
	Summary                  string `json:"summary"`
	PlantingChapter          string `json:"planting_chapter"`
	PlantingDescription      string `json:"planting_description"`
	Importance               string `json:"importance"`
	PlannedResolutionChapter string `json:"planned_resolution_chapter"`
}

// validForeshadowingTypes is the set of accepted ForeshadowingType values.
var validForeshadowingTypes = map[string]domain.ForeshadowingType{
	"hint":        domain.ForeshadowingTypeHint,
	"prophecy":    domain.ForeshadowingTypeProphecy,
	"mystery":     domain.ForeshadowingTypeMystery,
	"symbol":      domain.ForeshadowingTypeSymbol,
	"chekhov":     domain.ForeshadowingTypeChekhov,
	"red_herring": domain.ForeshadowingTypeRedHerring,
}

// validForeshadowingImportances is the set of accepted ForeshadowingImportance values.
var validForeshadowingImportances = map[string]domain.ForeshadowingImportance{
	"major":  domain.ForeshadowingImportanceMajor,
	"minor":  domain.ForeshadowingImportanceMinor,
	"subtle": domain.ForeshadowingImportanceSubtle,
}

// foreshadowingSlugRe matches characters that are not suitable in a slug ID.
var foreshadowingSlugRe = regexp.MustCompile(`[^a-zA-Z0-9_\-]`)

// foreshadowingSlugify converts a display name into a URL/file-safe identifier.
// Why: adopts the same simple snake_case convention used by the TypeScript
// side (spaces→underscores, lowercase, strip non-ascii).
func foreshadowingSlugify(name string) string {
	s := strings.ToLower(name)
	s = strings.ReplaceAll(s, " ", "_")
	s = foreshadowingSlugRe.ReplaceAllString(s, "")
	if s == "" {
		// Fallback for pure non-ASCII names (e.g. Japanese): use a hex digest
		// of the original to keep IDs stable and unique.
		s = fmt.Sprintf("f_%x", name)
	}
	return s
}

// Definition advertises the foreshadowing_create schema.
func (ForeshadowingCreateTool) Definition() protocol.Tool {
	return protocol.Tool{
		Name:        "foreshadowing_create",
		Description: "Create a new foreshadowing entity in the storyteller project",
		InputSchema: json.RawMessage(`{
			"type": "object",
			"properties": {
				"id":                        {"type": "string", "description": "Optional ID (slugified from name if omitted)"},
				"name":                      {"type": "string", "description": "Display name of the foreshadowing"},
				"type":                      {"type": "string", "enum": ["hint","prophecy","mystery","symbol","chekhov","red_herring"]},
				"summary":                   {"type": "string", "description": "Short description of the foreshadowing"},
				"planting_chapter":          {"type": "string", "description": "Chapter where the foreshadowing is planted"},
				"planting_description":      {"type": "string", "description": "How the foreshadowing is planted"},
				"importance":                {"type": "string", "enum": ["major","minor","subtle"]},
				"planned_resolution_chapter":{"type": "string", "description": "Chapter where resolution is planned"}
			},
			"required": ["name","type","summary","planting_chapter","planting_description"]
		}`),
	}
}

// Handle validates arguments and returns the created foreshadowing summary.
//
// Why: the tool intentionally does not persist to disk — consistent with
// the Phase-Green approach where MCP tools report the entity structure
// rather than writing files (that responsibility belongs to the CLI layer).
// The output confirms what would be created so the caller can verify inputs.
func (ForeshadowingCreateTool) Handle(_ context.Context, args json.RawMessage, _ ExecutionContext) (*protocol.CallToolResult, error) {
	var a foreshadowingCreateArgs
	if len(args) > 0 {
		_ = json.Unmarshal(args, &a)
	}

	// Validate required fields.
	if a.Name == "" {
		return &protocol.CallToolResult{
			Content: []protocol.ContentBlock{{Type: "text", Text: "name is required"}},
			IsError: true,
		}, nil
	}
	if a.Type == "" {
		return &protocol.CallToolResult{
			Content: []protocol.ContentBlock{{Type: "text", Text: "type is required"}},
			IsError: true,
		}, nil
	}
	if a.Summary == "" {
		return &protocol.CallToolResult{
			Content: []protocol.ContentBlock{{Type: "text", Text: "summary is required"}},
			IsError: true,
		}, nil
	}
	if a.PlantingChapter == "" {
		return &protocol.CallToolResult{
			Content: []protocol.ContentBlock{{Type: "text", Text: "planting_chapter is required"}},
			IsError: true,
		}, nil
	}
	if a.PlantingDescription == "" {
		return &protocol.CallToolResult{
			Content: []protocol.ContentBlock{{Type: "text", Text: "planting_description is required"}},
			IsError: true,
		}, nil
	}

	// Validate type enum.
	ft, ok := validForeshadowingTypes[a.Type]
	if !ok {
		return &protocol.CallToolResult{
			Content: []protocol.ContentBlock{{Type: "text", Text: fmt.Sprintf("invalid type %q: must be one of hint, prophecy, mystery, symbol, chekhov, red_herring", a.Type)}},
			IsError: true,
		}, nil
	}

	// Derive ID.
	id := a.ID
	if id == "" {
		id = foreshadowingSlugify(a.Name)
	}

	// Build the domain struct (in-memory only).
	f := &domain.Foreshadowing{
		ID:      id,
		Name:    a.Name,
		Type:    ft,
		Summary: a.Summary,
		Status:  domain.ForeshadowingStatusPlanted,
		Planting: domain.PlantingInfo{
			Chapter:     a.PlantingChapter,
			Description: a.PlantingDescription,
		},
	}

	if a.Importance != "" {
		if imp, ok := validForeshadowingImportances[a.Importance]; ok {
			f.Importance = &imp
		}
	}
	if a.PlannedResolutionChapter != "" {
		f.PlannedResolutionChapter = &a.PlannedResolutionChapter
	}

	// Serialize to JSON for the response body so callers can inspect the full struct.
	b, err := json.MarshalIndent(f, "", "  ")
	if err != nil {
		return &protocol.CallToolResult{
			Content: []protocol.ContentBlock{{Type: "text", Text: "internal error: " + err.Error()}},
			IsError: true,
		}, nil
	}

	text := fmt.Sprintf("foreshadowing created: %s\n%s", id, string(b))
	return &protocol.CallToolResult{
		Content: []protocol.ContentBlock{{Type: "text", Text: text}},
	}, nil
}
